### PR TITLE
Improve Fund Manager

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Match the operations of the FundMgr comprehensively with the expectations of the markets module.

The markets module is responsible for how much money is needed for deals in progress. The FundMgr now has a single responsibility: track how much is actually in the storage market actor for each address and add funds as needed when EnsureAvailable is called.

# Implementation

- Added a new state predicate for the StateChanged events API to track changes in market balances. This is comprised of StatePredicates.OnBalanceChanged and StatePredicate.AvailableBalanceChangedForAddresses, which compose with each other on top of StatePredicates.OnStorageMarketActorChanged. AvaialbleBalanceChangedForAddresses takes a function which returns an array of addresses to evaluate on each new change in tipset.
- Modified EnsureAvailable to:
   - read the available balance from state as needed
   - calculate the difference in available vs requested to determine if funds are added
   - pre-update available AHEAD of actually receiving a message that the chain message to add went through (if it fails will get reverted by the chain events monitor)
   - not "deduct" from available the amount to add.
- Add a setup function to setup monitoring of balances for the FundMgr
   - when balances change, always overwrite our local value (chain is authoritative)

# For Discussion

There is an edge case not covered here:

StartingTotal = 0,
EnsureAvailable(50)
EnsureAvailable(100)

will submit two messages to the chain, each for 50 (again, available updates ahead of the chain)

if the first submit fails, we end up with only 50 total. Should we track failures and resubmit? Probably. This is still an improvement over previous.

### CODE REVIEW

File: chain/events/state/predicates.go [hunk 1]
```diff
@@ -309,6 +354,57 @@ func (sp *StatePredicates) DealStateChangedForIDs(dealIds []abi.DealID) DiffAdtA
...
+			if err != nil {
+				return false, nil, err
+			}
+
+			newLockedBalance, err := newBalances.LockedTable.Get(addr)
+			if err != nil {
+				return false, nil, err
+			}
+
+			newBalance := big.Sub(newEscrowBalance, newLockedBalance)
```
Comment (whyrusleeping) [inline chain/events/state/predicates.go:395]: great shoes

**Comment by whyrusleeping** [comment]
Tests failing though.

**Comment by whyrusleeping** [comment]
> For Discussion

Yeah, seems like a case we want to handle, but the code here seems significantly improved, so lets do it as a followup

**Comment by hannahhoward** [comment]
@whyrusleeping should be resolved now

File: chain/market/fundmgr.go [hunk 2]
```diff
@@ -2,44 +2,123 @@ package market
...
+	return nil
+}
+
+// stateChanged handles balance changes monitored on the chain from one tipset to the next
+func (fm *FundMgr) stateChanged(ts *types.TipSet, ts2 *types.TipSet, states events.StateChange, h abi.ChainEpoch) (more bool, err error) {
+	changedBalances, ok := states.(state.ChangedBalances)
+	if !ok {
+		panic("Expected state.ChangedBalances")
+	}
+	// overwrite our in memory cache with new values from chain (chain is canonical)
```
Comment (magik6k) [inline chain/market/fundmgr.go:92]: We should probably keep track of in-flight AddFunds requests, probably by comparing sender nonce with nonces of messages we got from MpoolPushMessage below.

(Not blocking this PR)

**Pull request merged**